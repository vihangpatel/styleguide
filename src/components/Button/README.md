Button: Can choose between the button variants or the color of their choice
```js
  <Box row>
   <Box padding="5">
    <Button buttonVariant='default' isEnabled width={100} height={35} label='Default' showLoader/>
   </Box>
   <Box padding="5">
    <Button buttonVariant='primary' isEnabled width={100} height={35} label='Primary' />
    </Box>
   <Box padding="5">
    <Button buttonVariant='danger' isEnabled width={100} height={35} label='Danger' />
    </Box>
    <Box padding="5">
    <Button buttonVariant='success' isEnabled width={100} height={35} label='Success' />
   </Box>
   <Box padding="5">
    <Button buttonVariant='info' isEnabled width={100} height={35} label='Info' />
    </Box>
   <Box padding="5">
    <Button buttonVariant='accent' isEnabled width={100} height={35} label='Accent' />
    </Box>
  </Box>
```