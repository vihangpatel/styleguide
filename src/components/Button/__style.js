import theme from '../../utils/base-ui-theme'
import { darken } from 'polished'

/* *Style for the buttons */
export const __button = ({
	width,
	height,
	isEnabled,
	isLink,
	buttonVariant,
	isFlat,
	isTransparent,
	border,
	color,
}) => ({
	backgroundColor: `${
		isEnabled
			? theme.buttonVariants[buttonVariant] || buttonVariant
			: theme.buttonVariants.disabled
	}`,
	width: `${width}`,
	height: `${height}`,
	color: color || 'white',
	border: border || 'none',
	textTransform: 'capitalize',
	cursor: `${isEnabled ? 'pointer' : 'not-allowed'}`,
	borderRadius: theme.borderRadius,
	fontSize: 14,
	...theme.regularBaseFont,
	extend: [
		{
			condition: isEnabled && !isLink,
			style: {
				// onHover: {
				// 	backgroundColor: darken(
				// 		0.2,
				// 		theme.buttonVariants[buttonVariant] || buttonVariant
				// 	),
				// },
			},
		},
		{
			condition: isLink,
			style: {
				backgroundColor: theme.colorWhite,
				color: theme.buttonVariants[buttonVariant] || buttonVariant,
			},
		},
		{
			condition: isFlat,
			style: {
				backgroundColor: theme.colorWhite,
				borderRadius: 5,
				color: theme.color.mineShaft,
				border: 'solid 1px #d2d2d2',
				onHover: {
					backgroundColor: theme.colorWhite,
					border: 'solid 1px #d2d2d2',
				},
			},
		},
		{
			condition: isTransparent,
			style: {
				backgroundColor: 'initial',
				color: theme.color.mineShaft,
				opacity: '0.5',
				onHover: {
					backgroundColor: 'initial',
				},
			},
		},
	],

	onFocus: {
		outline: 'none',
	},
})
