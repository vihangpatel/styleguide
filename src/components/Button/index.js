import React from 'react'
import PropTypes from 'prop-types'
import { createComponent } from 'react-fela'
import { hrefHandler } from '../../utils/helpers'

/** Fetching style for the button component */
import { __button } from './__style'
import SubLoader from '../SubLoader/SubLoader'

/** creating elements using react-fela */
const ButtonBase = createComponent(__button, 'button')

const ButtonWrap = props => {
	const ComponentActive = (
		<ButtonBase
			{...props}
			onClick={hrefHandler(props.click)}
			passThrough={['onClick']}
		>
			{props.showLoader ? <SubLoader /> : props.label}
		</ButtonBase>
	)
	const ComponentBlocked = (
		<ButtonBase {...props}>
			{props.showLoader ? <SubLoader /> : props.label}
		</ButtonBase>
	)

	return props.isEnabled ? ComponentActive : ComponentBlocked
}

export default function Button({ ...props }) {
	return <ButtonWrap {...props} />
}

Button.propTypes = {
	/** Different Variants of Button component */
	buttonVariant: PropTypes.string,
	/** Callback to the listing page. It is a function */
	click: PropTypes.func,
	/** Disable or Enable a button */
	isEnabled: PropTypes.bool,
	/** Provide width of the button */
	width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	/** Provide height of the button */
	height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	/** Provide Button label */
	label: PropTypes.string,
	/** Show Loadering Text */
	showLoader: PropTypes.bool,
	/** Show flat theme. Used in Activity Tab switch */
	isFlat: PropTypes.bool,
}
