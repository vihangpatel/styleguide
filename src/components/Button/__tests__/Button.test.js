import React from 'react'
import Button from 'components/Button'
import renderer from 'react-test-renderer'
import StyleProvider from 'utils/Provider'
import { shallow, mount } from 'enzyme'

describe('Button Testing', () => {
	it('renders correctly', () => {
		const eventCard = renderer
			.create(
				<StyleProvider>
					<Button
						buttonVariant="default"
						isEnabled
						width={100}
						height={35}
						label="Default"
						showLoader
					/>
				</StyleProvider>
			)
			.toJSON()
		expect(eventCard).toMatchSnapshot()
	})

	it('renders correctly with enzyme', () => {
		const renderedComponent = shallow(
			<StyleProvider>
				<Button
					buttonVariant="default"
					isEnabled
					width={100}
					height={35}
					label="Default"
					showLoader
				/>
			</StyleProvider>
		)
		expect(renderedComponent.getElement('button')).toBeDefined()
	})

	it('should be clickable', () => {
		const clickSpy = jest.fn()
		const renderedComponent = mount(
			<StyleProvider>
				<Button
					buttonVariant="default"
					isEnabled
					width={100}
					height={35}
					label="Default"
					showLoader
					click={clickSpy}
				/>
			</StyleProvider>
		)
		renderedComponent.simulate('click')
		expect(clickSpy).toHaveBeenCalled()
	})
})
