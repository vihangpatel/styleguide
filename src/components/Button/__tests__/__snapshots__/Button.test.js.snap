// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Testing renders correctly 1`] = `
<button
  className="a b c d e f g h i j k l"
  onClick={[Function]}
>
  <div
    className="m n o p q r s"
  >
    <div
      className="t u v w x y z ab ac ae af ag ah ai aj ak al am an ao ap aq"
    />
    <div
      className="t u v w x ar z ab ac ae af ag ah ai aj ak al am an ao ap as"
    />
    <div
      className="t u v w x at z ab ac ae af ag ah ai aj ak al am an ao ap au"
    />
    <div
      className="t u v w x av z ab ac ae af ag ah ai aj ak al am an ao ap aw"
    />
    <div
      className="t u v w x ax z ab ac ae af ag ah ai aj ak al am an ao ap ay"
    />
    <div
      className="t u v w x az z ab ac ae af ag ah ai aj ak al am an ao ap ba"
    />
    <div
      className="t u v w x bb z ab ac ae af ag ah ai aj ak al am an ao ap bc"
    />
    <div
      className="t u v w x bd z ab ac ae af ag ah ai aj ak al am an ao ap be"
    />
    <div
      className="t u v w x bf z ab ac ae af ag ah ai aj ak al am an ao ap bg"
    />
    <div
      className="t u v w x bh z ab ac ae af ag ah ai aj ak al am an ao ap bi"
    />
    <div
      className="t u v w x bj z ab ac ae af ag ah ai aj ak al am an ao ap bk"
    />
    <div
      className="t u v w x bl z ab ac ae af ag ah ai aj ak al am an ao ap bm"
    />
  </div>
</button>
`;
